### PR TITLE
Group scheduler + code generation

### PR DESCRIPTION
**Pull request by amir734jj**
- `schedule_visits_group`: utility function that gets a group as an argument and tries to schedule all nodes in that group 
- `schedule_locals`: utility function that tries to schedule as many locals as possible
- `schedule_visits`: utility function that tries to schedule node that is possible to be scheduled
- `schedule_transitions`: utility function called when group scheduler finishes scheduling a group
- `assert_locals_order`: utility function that asserts that no local is scheduled out of order

### CODE REVIEW

**Review by boyland — COMMENTED** [review]
Not ready.  See comment in Teams.

File: analyze/aps-oag.c [hunk 1]
```diff
@@ -303,9 +305,298 @@ CTO_NODE* schedule_rest(AUG_GRAPH *aug_graph,
...
+ * Returns true if two attribute instances belong to the same group
+ * @param phase1 CHILD_PHASE* instance 1 
+ * @param phase2 CHILD_PHASE* instance 2
+ */ 
+static bool instances_are_in_same_group(CHILD_PHASE* group_key1, CHILD_PHASE* group_key2)
+{
+  // parent inherited attribute
+  if (group_key1->ph < 0 && group_key1->ch == -1 && group_key2->ph < 0 && group_key2->ch == -1)
+  {
+    return true;
```
Comment (boyland) [inline analyze/aps-oag.c:354]: But two inherited attributes of the parent in *different* phases are NOT in the same group.

File: analyze/aps-oag.c [hunk 2]
```diff
@@ -303,9 +305,298 @@ CTO_NODE* schedule_rest(AUG_GRAPH *aug_graph,
...
+  }
+  // parent synthesized attribute
+  else if (group_key1->ph > 0 && group_key1->ch == -1 && group_key2->ph > 0 && group_key2->ch == -1)
+  {
+    return true;
```
Comment (boyland) [inline analyze/aps-oag.c:359]: Similarly: two syn attributes of the parent from different phases are not in the same group.

File: analyze/aps-oag.c [hunk 3]
```diff
@@ -303,9 +305,298 @@ CTO_NODE* schedule_rest(AUG_GRAPH *aug_graph,
...
+  }
+  // child inherited attribute
+  else if (group_key1->ph < 0 && group_key1->ch > -1 && group_key2->ph < 0 && group_key2->ch > -1)
+  {
+    return true;
```
Comment (boyland) [inline analyze/aps-oag.c:364]: Ditto for child attributes. Even more so:  Two attributes that are both syn and both in the same phase but for *different* children are in different groups.

File: analyze/aps-oag.c [hunk 4]
```diff
@@ -303,9 +305,298 @@ CTO_NODE* schedule_rest(AUG_GRAPH *aug_graph,
...
+  }
+  // child inherited attribute
+  else if (group_key1->ph > 0 && group_key1->ch > -1 && group_key2->ph < 0 && group_key2->ch > -1)
+  {
+    return true;
+  }
+  // locals
+  else if (!group_key1->ph && !group_key1->ch && !group_key2->ph && !group_key2->ch)
+  {
+    return true;
```
Comment (boyland) [inline analyze/aps-oag.c:374]: Two locals are in the same group only if they are the same attribute instance.  Remember that locals do NOT group.

File: analyze/aps-oag.c [hunk 5]
```diff
@@ -303,9 +305,298 @@ CTO_NODE* schedule_rest(AUG_GRAPH *aug_graph,
...
+  
+  for (j = 0; j < n; j++)
+  {
+    // Already scheduled then ignore
+    if (aug_graph->schedule[j] != 0) continue;
+
+    // Instance in the same group but cannot be considered
+    if (instances_are_in_same_group(&instance_groups[i], &instance_groups[j]))
+    {
+      int index = j * n + i;    // i >--> j edge
```
Comment (boyland) [inline analyze/aps-oag.c:403]: I thought this code should were checking if j was also ready to go, but this code is only looking at dependencies between j and i.  "Ready to go" means all things it depends on are scheduled already.

File: analyze/aps-oag.c [hunk 6]
```diff
@@ -303,9 +305,298 @@ CTO_NODE* schedule_rest(AUG_GRAPH *aug_graph,
...
+ * @param cond current condition
+ * @param instance_groups array of <ph,ch>
+ * @param i instance index to test
+ */
+static bool local_ready_to_go(AUG_GRAPH *aug_graph, CONDITION cond, CHILD_PHASE* instance_groups, const int i)
+{
+  return true;
+}
+
+/**
```
Comment (boyland) [inline analyze/aps-oag.c:437]: You shouldn't need four different functions here.  Just use the CHILDPHASE

File: analyze/aps-oag.c [hunk 7]
```diff
@@ -332,25 +623,52 @@ void schedule_augmented_dependency_graph(AUG_GRAPH *aug_graph) {
...
 	   aug_graph_name(aug_graph));
   }
-  if (oag_debug & DEBUG_ORDER) {
-    for (int i=0; i <= n; ++i) {
+
+  size_t instance_groups_size = n * sizeof(CHILD_PHASE);
+  CHILD_PHASE* instance_groups = (CHILD_PHASE*) alloca(instance_groups_size);
+  memset(instance_groups, (int)0, instance_groups_size);
+
+  for (i = 0; i < n; i++)
```
Comment (boyland) [inline analyze/aps-oag.c:631]: Looks good.

File: analyze/aps-oag.c [hunk 8]
```diff
@@ -303,9 +305,298 @@ CTO_NODE* schedule_rest(AUG_GRAPH *aug_graph,
...
+    CHILD_PHASE instance_group = instance_groups[i];
+
+    // Already scheduled then ignore
+    if (aug_graph->schedule[i] != 0) continue;
+
+    /* check to see if makes sense
+    * (No need to schedule something that
+    * occurs only in a different condition branch.)
+    */
+    if (MERGED_CONDITION_IS_IMPOSSIBLE(cond, instance_condition(instance))) return false;
```
Comment (boyland) [inline analyze/aps-oag.c:561]: Not sure if "false" is the right thing.  or maybe should schedule.  WHat is the purpose of scheduling an impossible instance?

File: analyze/aps-oag.c [hunk 9]
```diff
@@ -303,9 +305,298 @@ CTO_NODE* schedule_rest(AUG_GRAPH *aug_graph,
...
+
+    // If edgeset condition is not impossible then go ahead with scheduling
+    if (generic_instance_ready_to_go(aug_graph, cond, instance_groups, i))
```
Comment (boyland) [inline analyze/aps-oag.c:564]: Looks like this should work.  Maybe even: should go ahead and schedule *all* inthe group if they are all ready to go together.

**Comment by amir734jj** [comment]
This PR is very outdated. Closing

**Pull request closed without merging**